fedimint-api/src/tiered_multi.rs

```diff
@@ -141,19 +142,42 @@ where
 }
 
 impl TieredMulti<()> {
-    // TODO: move somewhere else?
-    pub fn represent_amount<K>(mut amount: Amount, tiers: &Tiered<K>) -> TieredMulti<()> {
-        let coins = tiers
-            .tiers()
-            .rev()
-            .map(|&amount_tier| {
-                let res = amount / amount_tier;
-                amount %= amount_tier;
-                (amount_tier, vec![(); res as usize])
-            })
-            .collect();
+    /// Determines the denominations to use when representing an amount
+    ///
+    /// Algorithm tries to leave the user with a target number of `denomination_sets` starting
+    /// at the lowest denomination.  `self` gives the denominations that the user already has.
+    pub fn represent_amount<K, V>(
+        mut amount: Amount,
+        current_denominations: &TieredMulti<V>,
+        tiers: &Tiered<K>,
+        denomination_sets: u16,
+    ) -> TieredMulti<()> {
+        let mut denominations: Vec<(Amount, ())> = vec![];
+        let mut tiers: Vec<Amount> = tiers.tiers().cloned().collect();
+        tiers.sort();
+
+        // try to hit the target `denomination_sets`
+        for tier in tiers.clone() {
+            let notes = current_denominations
+                .get(tier)
+                .map(|v| v.len())
+                .unwrap_or(0);
+            let missing_notes = (denomination_sets as u64).saturating_sub(notes as u64);
+            let possible_notes = amount / tier;
+
+            let add_notes = min(possible_notes, missing_notes);
+            denominations.extend(vec![(tier, ()); add_notes as usize]);
+            amount -= tier * add_notes;
+        }
+
+        // if there is a remaining amount, add denominations with a greedy algorithm
+        for tier in tiers.into_iter().rev() {
+            let res = amount / tier;
+            amount %= tier;
+            denominations.extend(vec![(tier, ()); res as usize]);
+        }
 
-        TieredMulti(coins)
+        denominations.into_iter().collect()
```

Comment:
Can we put an assert here that we actually represent the right amount (just in case)?